src/NHibernate/Linq/LinqExtensionMethodAttribute.cs

```diff
@@ -2,17 +2,103 @@
 
 namespace NHibernate.Linq
 {
-	public class LinqExtensionMethodAttribute: Attribute
+	/// <summary>
+	/// Flag a method as being a SQL function call for the linq-to-nhibernate provider. Its
+	/// parameters will be used as the function call parameters.
+	/// </summary>
+	public class LinqExtensionMethodAttribute : LinqExtensionMethodAttributeBase
 	{
+		/// <summary>
+		/// Default constructor. The method call will be translated by the linq provider to
+		/// a function call having the same name than the method.
+		/// </summary>
 		public LinqExtensionMethodAttribute()
-		{
-		}
+			: base(LinqExtensionEvaluation.NoEvaluation) { }
 
+		/// <summary>
+		/// Constructor specifying a SQL function name.
+		/// </summary>
+		/// <param name="name">The name of the SQL to call.</param>
 		public LinqExtensionMethodAttribute(string name)
+			: this(name, LinqExtensionEvaluation.NoEvaluation) { }
+
+		/// <summary>
+		/// Constructor allowing to specify a <see cref="LinqExtensionEvaluation"/> for the method.
+		/// </summary>
+		/// <param name="evaluation">Should the method call be partially evaluated? Default is
+		/// <see cref="LinqExtensionEvaluation.NoEvaluation"/>.</param>
+		public LinqExtensionMethodAttribute(LinqExtensionEvaluation evaluation)
+			: base(evaluation) { }
+
+		/// <summary>
+		/// Constructor for specifying a SQL function name and a <see cref="LinqExtensionEvaluation"/>.
+		/// </summary>
+		/// <param name="name">The name of the SQL function to call.</param>
+		/// <param name="evaluation">Should the method call be partially evaluated? Default is
+		/// <see cref="LinqExtensionEvaluation.NoEvaluation"/>.</param>
+		public LinqExtensionMethodAttribute(string name, LinqExtensionEvaluation evaluation)
+			: base(evaluation)
 		{
 			Name = name;
 		}
 
-		public string Name { get; private set; }
+		/// <summary>
+		/// The name of the SQL to call.
+		/// </summary>
+		public string Name { get; }
+	}
+
+	/// <summary>
+	/// Can flag a method as not being callable by the runtime, when used in Linq queries.
+	/// If the method is supported by the linq-to-nhibernate provider, it will always be converted
+	/// to the corresponding SQL statement.
+	/// Otherwise the linq-to-nhibernate provider evaluates method calls when they do not depend on
+	/// the queried data.
+	/// </summary>
+	public class NotEvaluatableAttribute : LinqExtensionMethodAttributeBase
+	{
+		/// <summary>
+		/// Default constructor.
+		/// </summary>
+		public NotEvaluatableAttribute()
+			: base(LinqExtensionEvaluation.NoEvaluation) { }
+	}
+
+	/// <summary>
+	/// Base class for Linq extension attributes.
+	/// </summary>
+	public abstract class LinqExtensionMethodAttributeBase : Attribute
+	{
+		/// <summary>
+		/// Should the method call be partially evaluated?
+		/// </summary>
+		public LinqExtensionEvaluation Evaluation { get; }
```

Comment:
Talking about evaluation here is to disclose the internals of the implementation, which do not matter in this context. This needs to be close to the context. In this context if something is not evaluatable it will be forcible translated into a method call; if it is partially evaluatable, then it will be calculated before the call. I don't know how to express this as a name of the property.